### DISCUSSION TITLE
Remaining Clang-Tidy Checks

### DISCUSSION THREAD

**Discussion by Noordfrees** [widelands/widelands#5262] [Show and tell]
Following the discussion in #5246, I've made a list of checks that are not yet being enforced by our clang-tidy CI. I've categorized them as:
- Perm-suppressed: We really need to keep these hits around for various reasons.
- Improves performance or code safety
- Helps to detect and guard against bugs
- Possibly useful stylistic hints; improves readability
- Undesired stuff nobody needs or wants

This categorization reflects my personal opinion and is open for discussion.

---

# Perm-suppressed checks

- warning: do not define a C-style variadic function; consider using a function parameter pack or currying instead [cert-dcl50-cpp]
- warning: do not call c-style vararg functions [cppcoreguidelines-pro-type-vararg,hicpp-vararg]

# Improves performance

- warning: pass by value and use std::move [modernize-pass-by-value]
- warning: parameter 'n' is passed by value and only copied once; consider moving it to avoid unnecessary copies [performance-unnecessary-value-param]
- warning: method 'shutdown_settings' can be made static [readability-convert-member-functions-to-static]

# Helps against bugs

- warning: an exception may be thrown in function '~BufferedConnection' which should not throw exceptions [bugprone-exception-escape]
- warning: macro argument should be enclosed in parentheses [bugprone-macro-parentheses]
- warning: narrowing conversion from 'int64_t' (aka 'long') to signed type 'int' is implementation-defined [bugprone-narrowing-conversions,cppcoreguidelines-narrowing-conversions]
- warning: comparison length is too long and might lead to a buffer overflow [bugprone-not-null-terminated-result]
- warning: 'signed char' to 'const int16_t' (aka 'const short') conversion; consider casting to 'unsigned char' first. [bugprone-signed-char-misuse,cert-str34-c]
- warning: loop variable has narrower type 'uint8_t' (aka 'unsigned char') than iteration\'s upper bound 'uint32_t' (aka 'unsigned int') [bugprone-too-small-loop-variable]
- warning: initialization of 'wl_bid' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp]
- warning: Called C++ object pointer is null [clang-analyzer-core.CallAndMessage]
- warning: Forming reference to null pointer [clang-analyzer-core.NonNullParamChecker]
- warning: The left operand of '<' is a garbage value [clang-analyzer-core.UndefinedBinaryOperatorResult]
- warning: Use of memory after it is freed [clang-analyzer-cplusplus.NewDelete]
- warning: Potential memory leak [clang-analyzer-cplusplus.NewDeleteLeaks]
- warning: 2 uninitialized fields at the end of the constructor call [clang-analyzer-optin.cplusplus.UninitializedObject]
- warning: Call to virtual method 'Flag::init' during construction bypasses virtual dispatch [clang-analyzer-optin.cplusplus.VirtualCall]
- warning: variable 'end' is not initialized [cppcoreguidelines-init-variables]
- warning: uninitialized record type: 'tfile' [cppcoreguidelines-pro-type-member-init,hicpp-member-init]
- warning: slicing object from type 'FCoords' to 'Coords' discards 12 bytes of state [cppcoreguidelines-slicing]
- warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
- warning: return type 'const std::vector<std::string>' (aka 'const vector<basic_string<char>>') is 'const'-qualified at the top level, which may reduce code readability without improving const correctness [readability-const-return-type]

# Possibly useful style stuff

- warning: 60 is a magic number; consider replacing it with a named constant [cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers]
- warning: function-like macro 'FP' used; consider a 'constexpr' template function [cppcoreguidelines-macro-usage]
- warning: class 'GameLogicThread' defines a non-default destructor, a copy constructor and a copy assignment operator but does not define a move constructor or a move assignment operator [cppcoreguidelines-special-member-functions,hicpp-special-member-functions]
- warning: function 'construct_building' exceeds recommended size/complexity thresholds [google-readability-function-size,hicpp-function-size,readability-function-size]
- warning: switch could be better written as an if/else statement [hicpp-multiway-paths-covered]
- warning: all parameters should be named in a function [hicpp-named-parameter,readability-named-parameter]

# Undesired nonsense

- warning: do not declare C-style arrays, use std::array<> instead [cppcoreguidelines-avoid-c-arrays,hicpp-avoid-c-arrays,modernize-avoid-c-arrays]
- warning: avoid using 'goto' for flow control [cppcoreguidelines-avoid-goto,hicpp-avoid-goto]
- warning: do not manage memory manually; use RAII [cppcoreguidelines-no-malloc,hicpp-no-malloc]
- warning: member variable 'w_' has protected visibility [cppcoreguidelines-non-private-member-variables-in-classes,misc-non-private-member-variables-in-classes]
- warning: deleting a pointer through a type that is not marked 'gsl::owner<>'; consider using a smart pointer instead [cppcoreguidelines-owning-memory]
- warning: do not implicitly decay an array into a pointer; consider using gsl::array_view or an explicit cast instead [cppcoreguidelines-pro-bounds-array-to-pointer-decay,hicpp-no-array-decay]
- warning: do not use array subscript when the index is not an integer constant expression; use gsl::at() instead [cppcoreguidelines-pro-bounds-constant-array-index]
- warning: do not use pointer arithmetic [cppcoreguidelines-pro-bounds-pointer-arithmetic]
- warning: do not use const_cast [cppcoreguidelines-pro-type-const-cast]
- warning: do not use reinterpret_cast [cppcoreguidelines-pro-type-reinterpret-cast]
- warning: do not access members of unions; use \(boost::)variant instead [cppcoreguidelines-pro-type-union-access]
- warning: default arguments on virtual or override methods are prohibited [google-default-arguments]
- warning: floating point literal has suffix 'f', which is not uppercase [hicpp-uppercase-literal-suffix,readability-uppercase-literal-suffix]
- warning: use auto when initializing with a cast to avoid duplicating the type name [hicpp-use-auto,modernize-use-auto]
- warning: 'operator<<<std::char_traits<char>>' must resolve to a function declared within the '__llvm_libc' namespace [llvmlibc-callee-namespace]
- warning: escaped string literal can be written as a raw string literal [modernize-raw-string-literal]
- warning: avoid repeating the return type from the declaration; use a braced initializer list instead [modernize-return-braced-init-list]
- warning: use a trailing return type for this function [modernize-use-trailing-return-type]

**Comment by hessenfarmer** [comment]
@Noordfrees Do I understand this correct that this list is based on the warnings thrown by our current code?        
Could there be other useful checks in clang-tidy that aren't just triggered by luck or the experience of all our coders so far.    
How many and which checks are currently activated?     
In the list I see there are some checks performed by clang-analyzer rather then clang-tidy. is this invoked automatically by clang-tidy or do we need to trigger this in our GitHub action?    
  
Sorry for posing so many questions, but I just want to understand what we are doing here.    
@widelands/developers any opinions?

**Reply by Noordfrees** [reply]
Our clang-tidy results analyser works on an exclusion principle, that is, any clang-tidy hit that is not explicitly listed in our [suppression list](https://github.com/widelands/widelands/blob/master/utils/check_clang_tidy_results.py#L11) causes the CI to go red. The list in the opening post only includes checks we are actively suppressing at the moment; all the hundreds of other checks clang-tidy can do are already clean and are being checked against in the CI.

There are a few exceptions though, the 22 checks `android*` and `fuchsia*` are excluded completely. I have not yet looked at what those checks are about and whether there are any among them we might want to clear as well.

And questions are of course always welcome :)

**Reply by jmoerschbach** [reply]
Actually I trust the maintainers of clang-tidy/clang-analyzer for providing rational rules and am in favor of turning on as many as possible. I think Nordfrees did a good job already sorting out the ones which make not much of sense to this project :)

**Comment by hessenfarmer** [comment]
Well from my professional life I am used to first define a coding standard for a project and afterwards enforcing it with a sufficient tool. I can adopt to choose the tool and then enforce the checks it provides.    
However in this case we should probably enforce all of them that are applicable for our code base rather then choosing them on a subjective basis.
From my point of view we could safely exclude the following groups:
abseil - we do not use this library
altera - we are not using OpenCL
android - we are not supporting android (yet)
boost - we do not use it anymore
fuchsia -  we do not support this OS
linuxkernel - obviously we are not programming any kernel
Zircon - same as linuxkernel
this may safe some execution time of the check perhaps.

each other check would be sensible to activate imho.  
For example I checked for the usage of malloc and only found a couple of uses in our code but one of it had already a TODO comment about a memory leak.

**Comment by tothxa** [comment]
I can't comment on which tests to include or exclude, but when the list is finalised, please create a `.clang-tidy` config file, instead of supplying them each time to `run-clang-tidy.py`.

Also currently `HeaderFilterRegex`/`-header-filter` is unset, so headers are actually not checked. Setting it to `'.*'` produces lots of errors. (that's in project headers, as `-system-headers` is a separate option, off by default) I guess those should also be cleaned up.

I'll still venture a question regarding @Noordfrees's perm-suppressed list: These look quite scary, and I think they're only used in a few critical places. Shouldn't the critical parts get overrides instead of ignoring the checks completely?

**Comment by hessenfarmer** [comment]
well the list is not up to date maybe we could date it up

Edit: cleaned up and updated the list
